server/service/jct/api.coffee

```diff
@@ -787,51 +783,54 @@ API.service.jct.permission = (issn, institution) ->
 
 
 # Calculate self archiving check. It combines, sa_prohibited, OA.works permission and rr checks
-API.service.jct.sa = (journal, institution, funder, retention=true, sa_prohibition=true) ->
-  # Get SA prohibition
-  if journal and sa_prohibition
-    res_sa = API.service.jct.sa_prohibited journal, undefined
-    if res_sa and res_sa.compliant is 'no'
-      return res_sa
+API.service.jct.sa = (journal, institution, funder) ->
+
+  _merge_logs_and_qa= (res1, res2) ->
+    # merge the qualifications and logs from res1 into res2
+    res2.qualifications ?= []
+    if res1.qualifications? and res1.qualifications.length
+      for q in (if _.isArray(res1.qualifications) then res1.qualifications else [res1.qualifications])
+        res2.qualifications.push(q)
+    res2.log ?= []
+    if res1.log? and res1.log.length
+      for l in (if _.isArray(res1.log) then res1.log else [res1.log])
+        res2.log.push(l)
+    return res2
+
+  # Check SA prohibition
+  res_sa = API.service.jct.sa_prohibited journal, undefined
+  if res_sa and res_sa.compliant is 'no'
+    return res_sa
+
+  # Get rights retention data
+  res_r = API.service.jct.retention journal, undefined
+  # merge the qualifications and logs from SA prohibition
+  res_r = _merge_logs_and_qa(res_sa, res_r)
+  if res_r.compliant is 'yes'
+    return res_r
 
   # Get OA.Works permission
   rs = API.service.jct.permission journal, institution
 
-  # merge the qualifications and logs from SA prohibition into OA.Works permission
-  rs.qualifications ?= []
-  if res_sa.qualifications? and res_sa.qualifications.length
-    for q in (if _.isArray(res_sa.qualifications) then res_sa.qualifications else [res_sa.qualifications])
-      rs.qualifications.push(q)
-  rs.log ?= []
-  if res_sa.log? and res_sa.log.length
-    for l in (if _.isArray(res_sa.log) then res_sa.log else [res_sa.log])
-      rs.log.push(l)
+  # merge the qualifications and logs from SA prohibition and SA rights retention into OA.Works permission
+  rs = _merge_logs_and_qa(res_r, rs)
 
   # check for retention
   if rs
     _rtn = {}
     for r in (if _.isArray(rs) then rs else [rs])
```

Comment:
I'm struggling to follow the logic in this section, but looking at the workflow diagram it seems like we can make a shortcut here in that if the OAB permissions check comes back as compliant, then we can return the result straight away, without going through this retention section first

Reply:
Discussed and understood current code